src/rgw/rgw_notify.cc

```diff
@@ -2,17 +2,466 @@
+      if (remove_entries) {
+        librados::ObjectWriteOperation delete_op;
+        cls_2pc_queue_remove_entries(delete_op, end_marker); 
+        ret = rgw_rados_operate(rados_ioctx, queue_name, &delete_op, optional_yield(io_context, yield)); 
+        if (ret < 0) {
+          ldout(cct, 1) << "ERROR: failed to remove entries up to: " << end_marker <<  " from queue: " 
+            << queue_name << ". error: " << ret << dendl;
+        } else {
+          ldout(cct, 20) << "INFO: removed entries up to: " << end_marker <<  " from queue: " 
+          << queue_name << dendl;
+        }
+      }
+      start_marker = end_marker;
+      // TODO: cleanup expired reservations
+    }
+  }
+
+  // process all queues
+  // find which of the queues is owned by this daemon and process it
+  void process_queues(spawn::yield_context yield) {
+    auto has_error = false;
+    while (true) {
+      Timer timer(io_context);
+      const auto duration = has_error ? 
+        std::chrono::milliseconds(queues_update_retry_ms) : std::chrono::milliseconds(queues_update_period_ms);
+      timer.expires_from_now(duration);
+      const auto tp = ceph::coarse_real_time::clock::to_time_t(ceph::coarse_real_time::clock::now() + duration);
+      ldout(cct, 20) << "INFO: next queue processing will happen at: " << std::ctime(&tp)  << dendl;
+      boost::system::error_code ec;
+      timer.async_wait(yield[ec]);
+
+      queues_t queues;
+      auto ret = read_queue_list(queues, optional_yield(io_context, yield));
+      if (ret < 0) {
+        has_error = true;
+        continue;
+      }
+
+      for (const auto& queue_name : queues) {
+        // try to lock the queue to check if it is owned by this rgw
+        // or if ownershif needs to be taken
+        librados::ObjectWriteOperation op;
+        rados::cls::lock::lock(&op, queue_name+"_lock", 
+              LOCK_EXCLUSIVE,
+              lock_cookie, 
+              "" /*no tag*/,
+              "" /*no description*/,
+              failover_time,
+              LOCK_FLAG_MAY_RENEW);
+
+        ret = rgw_rados_operate(rados_ioctx, queue_name, &op, optional_yield(io_context, yield));
+        if (ret == -EBUSY) {
+          // lock is already taken by another RGW
+          ldout(cct, 20) << "INFO: queue: " << queue_name << " owned (locked) by another daemon" << dendl;
+          continue;
+        }
+        if (ret < 0) {
+          // failed to lock for another reason, continue to process other queues
+          ldout(cct, 1) << "ERROR: failed renew lock on queue: " << queue_name << ". error: " << ret << dendl;
+          has_error = true;
+        }
+        // add queue to list of owned queues
+        if (owned_queues.insert(queue_name).second) {
+          ldout(cct, 20) << "INFO: queue: " << queue_name << " now owned (locked) by this daemon" << dendl;
+          // start processing this queue
+          spawn::spawn(io_context, [this, queue_name](spawn::yield_context yield) {
```

Comment:
`spawn(yield,` here to share the strand executor

Reply:
processing should be serialized within a queue, isn't spawn(yield...) in line 197 achieving that?
among queues, I would prefer them being multiplexed (or executed in parallel in case of more than 1 thread)

Reply:
ok, as long as separate invocations of process_queue() are thread-safe

Reply:
you are right, they all touch the `owned_queue` list (erase from it).
will change so that only `process_queues()` erase from the list

Reply:
yeah, io_context and rados_ioctx are thread-safe

Reply:
got it. thanks